### ISSUE TITLE
Container ID inconsistency results in deleted vlan device

### ISSUE THREAD

**Issue by JacobHenner** [aws/amazon-vpc-cni-k8s#1644]
**What happened**:

On several occasions, we've encountered an issue where pods scheduled on a node start with broken network connectivity and never recover. The containers within the pod will attempt to start repeatedly, but since the applications within depend on network connectivity, the containers exit in failure each time and the pod remains in CrashLoopBackoff.

Each time I've observed this personally, I've noticed it seems to be associated with many pods being scheduled on the node simultaneously. All of the pods scheduled around the same time have the same issue. However, the issue is not permanent - pods scheduled sometime after the issue has occurred start with functional network connectivity, and continue to work as expected.

After some investigation, I believe I've identified the immediate cause of the broken network connectivity. However, I'm unsure of the root cause, and whether the vpc-cni should be preventing the condition from occurring.

Comparing a healthy pod and an unhealthy pod from the same node, I noticed that the unhealthy pod does not have a vlan device (e.g. `vlan.eth.XY@eth1`) on the host, and that the associated routes that would otherwise be in place are missing. There is a veth device for the broken pod (e.g. `vlan7cf3df5c6e2@if3`).

Exploring this further, I found the following in the CNI logs (for a broken pod):

```jsonl
{"level":"info","ts":"2021-09-24T22:52:24.230-0400","caller":"routed-eni-cni-plugin/cni.go:117","msg":"Received CNI add request: ContainerID(30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c) Netns(/proc/20683/ns/net) IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=redacted;K8S_POD_NAME=redacted-bf4687b75-bsmsm;K8S_POD_INFRA_CONTAINER_ID=30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}

{"level":"info","ts":"2021-09-24T22:52:24.287-0400","caller":"routed-eni-cni-plugin/cni.go:117","msg":"Received add network response for container 30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c interface eth0: Success:true IPv4Addr:\"100.64.67.106\" DeviceNumber:-1 VPCcidrs:\"10.12.0.0/17\" VPCcidrs:\"100.64.0.0/17\" VPCcidrs:\"100.64.128.0/18\" PodVlanId:28 PodENIMAC:\"0e:01:d3:36:cc:57\" PodENISubnetGW:\"100.64.64.1\" ParentIfIndex:3 "}

{"level":"info","ts":"2021-09-24T22:52:26.955-0400","caller":"routed-eni-cni-plugin/cni.go:249","msg":"Received CNI del request: ContainerID(a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a) Netns() IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=redacted;K8S_POD_NAME=redacted-bf4687b75-bsmsm;K8S_POD_INFRA_CONTAINER_ID=a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}

{"level":"info","ts":"2021-09-24T22:52:26.965-0400","caller":"routed-eni-cni-plugin/cni.go:249","msg":"Received del network response for pod redacted-bf4687b75-bsmsm namespace redacted sandbox a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a: Success:true IPv4Addr:\"100.64.67.106\" PodVlanId:28 "}
```

From this, my suspicion is that the pod sandbox (e.g. the pause container) changed (from a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a to 30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c), and that the CNI deleted the interface configuration for the _pod_ when it got the CNI del request for the a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a _container_, even though 30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c was still running.

The kubelet logs for those container IDs are as follows:

```
Sep 24 22:52:08 ip-100-64-91-119.ec2.internal kubelet[5518]: W0924 22:52:08.303127    5518 helpers.go:291] Unable to create pod sandbox due to conflict. Attempting to remove sandbox "a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a"
Sep 24 22:52:08 ip-100-64-91-119.ec2.internal kubelet[5518]: E0924 22:52:08.303980    5518 helpers.go:296] Failed to remove the conflicting container "a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a": Error: No such container: a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a
Sep 24 22:52:26 ip-100-64-91-119.ec2.internal kubelet[5518]: W0924 22:52:26.628642    5518 docker_sandbox.go:402] failed to read pod IP from plugin/docker: networkPlugin cni failed on the status hook for pod "redacted-bf4687b75-bsmsm_redacted": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a"
Sep 24 22:52:26 ip-100-64-91-119.ec2.internal kubelet[5518]: I0924 22:52:26.639409    5518 kubelet.go:1899] SyncLoop (PLEG): "redacted-bf4687b75-bsmsm_redacted(012d1b8d-aaf5-432d-a45f-d2273aa8f6bc)", event: &pleg.PodLifecycleEvent{ID:"012d1b8d-aaf5-432d-a45f-d2273aa8f6bc", Type:"ContainerDied", Data:"30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c"}
Sep 24 22:52:26 ip-100-64-91-119.ec2.internal kubelet[5518]: W0924 22:52:26.639534    5518 pod_container_deletor.go:79] Container "30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c" not found in pod's containers
Sep 24 22:52:26 ip-100-64-91-119.ec2.internal kubelet[5518]: I0924 22:52:26.639560    5518 kubelet.go:1899] SyncLoop (PLEG): "redacted-bf4687b75-bsmsm_redacted(012d1b8d-aaf5-432d-a45f-d2273aa8f6bc)", event: &pleg.PodLifecycleEvent{ID:"012d1b8d-aaf5-432d-a45f-d2273aa8f6bc", Type:"ContainerDied", Data:"a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a"}
Sep 24 22:52:26 ip-100-64-91-119.ec2.internal kubelet[5518]: W0924 22:52:26.639781    5518 pod_container_deletor.go:79] Container "a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a" not found in pod's containers
Sep 24 22:52:26 ip-100-64-91-119.ec2.internal kubelet[5518]: W0924 22:52:26.930676    5518 cni.go:333] CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "a279fbfcc13e5c1f951b2b770b196f4cc4ac1630b80efc79707ca509d51e079a"
Sep 24 22:52:28 ip-100-64-91-119.ec2.internal kubelet[5518]: I0924 22:52:28.378542    5518 kubelet.go:1899] SyncLoop (PLEG): "redacted-bf4687b75-bsmsm_redacted(012d1b8d-aaf5-432d-a45f-d2273aa8f6bc)", event: &pleg.PodLifecycleEvent{ID:"012d1b8d-aaf5-432d-a45f-d2273aa8f6bc", Type:"ContainerStarted", Data:"30c860ea0116032d24b70be358ce519ca613b2a449a2ffaafd31992c3245359c"}
```

I'm not sure what to make of these logs. I don't imagine these failures are expected, but I do imagine they're related.

Looking at the docker daemon logs for the same time, I see:

```
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:22.307003534-04:00" level=error msg="Handler for GET /v1.40/containers/74f181080a27d51e82642ed53520d02b6a193fc1c77eaf0732170cbf4d059df2/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:22.307310668-04:00" level=error msg="Handler for GET /v1.40/containers/74f181080a27d51e82642ed53520d02b6a193fc1c77eaf0732170cbf4d059df2/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:22.307317248-04:00" level=error msg="Handler for GET /v1.40/containers/74f181080a27d51e82642ed53520d02b6a193fc1c77eaf0732170cbf4d059df2/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: http: superfluous response.WriteHeader call from github.com/docker/docker/api/server/httputils.WriteJSON (httputils_write_json.go:11)
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:22.307491880-04:00" level=error msg="Handler for GET /v1.40/containers/74f181080a27d51e82642ed53520d02b6a193fc1c77eaf0732170cbf4d059df2/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: http: superfluous response.WriteHeader call from github.com/docker/docker/api/server/httputils.WriteJSON (httputils_write_json.go:11)
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: http: superfluous response.WriteHeader call from github.com/docker/docker/api/server/httputils.WriteJSON (httputils_write_json.go:11)
Sep 24 22:52:22 ip-100-64-91-119.ec2.internal dockerd[5396]: http: superfluous response.WriteHeader call from github.com/docker/docker/api/server/httputils.WriteJSON (httputils_write_json.go:11)
Sep 24 22:52:26 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:26.468708964-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:42 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:42.565667606-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:44 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:44.673231937-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:44 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:44.756816200-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:44 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:44.849625554-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:51 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:51.480661606-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:51 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:51.757306502-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:52 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:52.170274288-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:52 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:52.255186091-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:56 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:56.172602355-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:56 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:56.268824056-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:56 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:56.455027055-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:56 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:56.553398051-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:56 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:56.870321436-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Sep 24 22:52:57 ip-100-64-91-119.ec2.internal dockerd[5396]: time="2021-09-24T22:52:57.166208893-04:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
```

I am not sure what to make of these either, but I do know that `74f181080a27d51e82642ed53520d02b6a193fc1c77eaf0732170cbf4d059df2` is the container ID of aws-vpc-cni .

My theory is that there's some inconsistency between dockerd, the kubelet, and the CNI plugin, which results in the CNI deleting the vlan interface that's needed by a running pod. I'm not sure what underlying condition triggers the inconsistency, but I imagine the CNI should do everything it can to avoid deleting an in-use interface (or at least replace it if it discovers it was deleted in error).

**Attach logs**

These logs are available in the AWS support case (see below).

**What you expected to happen**:

I expected pods to start with functional network connectivity.

**How to reproduce it (as minimally and precisely as possible)**:

I do not have a reproducer at this time, but will attempt to devise one as time permits.

**Anything else we need to know?**:

AWS support case 8957900351 was opened for this issue.

**Environment**:
- Kubernetes version (use `kubectl version`): `Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.20-eks-8c579e", GitCommit:"8c579edfc914f013ff48b2a2b2c1308fdcacc53f", GitTreeState:"clean", BuildDate:"2021-07-31T01:34:13Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}`
- CNI Version: 1.7.9
- OS (e.g: `cat /etc/os-release`): 

```
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"
```

- Kernel (e.g. `uname -a`): `Linux ip-100-64-91-119.ec2.internal 5.10.59-52.142.amzn2.x86_64 #1 SMP Wed Aug 18 21:17:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by chlunde** [comment]
Hi, we see a similar issue (w/ CNI 1.9.1 and bottlerocket 1.2.1, securitygroups per pod). Our issue happens with the prometheus statefulset pods, and we did not notice it until after we added storage for it. Do you also use statefulsets?

**Comment by chlunde** [comment]
It looks like this code deletes the interface based on name/namespace only. I think this could mean that if a delete event arrives after a pod name (and veth name) has been reused, it will tear down the CNI setup which is used by another instance:

https://github.com/aws/amazon-vpc-cni-k8s/blob/07284393715b6e3872b91a6eb94da803af4ab86e/cmd/routed-eni-cni-plugin/cni.go#L194-L199

**Comment by SaranBalaji90** [comment]
@JacobHenner @chlunde Thank you for your initial investigation. The problem here could be if previous pod network is not cleaned up completely and when DEL for old pod comes in after new pod is provisioned then it could delete incorrect network. We are working on mitigating this by adding local store to clean up appropriate VLAN network when DEL request is received. In the meanwhile, are looking into other alternatives to see if we can temporarily mitigate the impact while file based deletion approach is launched.

**Comment by chlunde** [comment]
@SaranBalaji90 thanks for looking into this!

If I understand correctly you are talking about a bigger change as a long term fix.

Here is a _draft_ of the mitigation I was thinking of which is localized to a single function. I think would greatly reduce the likelihood of this issue. It would also reduce the likelihood of having to wait for cleanup if proper "collision detection" is implemented.

```patch
diff --git a/pkg/provider/branch/trunk/trunk.go b/pkg/provider/branch/trunk/trunk.go
index 803e789..6a7e498 100644
--- a/pkg/provider/branch/trunk/trunk.go
+++ b/pkg/provider/branch/trunk/trunk.go
@@ -98,6 +98,8 @@ type trunkENI struct {
        instance ec2.EC2Instance
        // usedVlanIds is the list of boolean value representing the used vlan ids
        usedVlanIds []bool
+       // lastVlanIdx is the index of the most recently allocated VLAN ID
+       lastVlanIdx int
        // branchENIs is the list of BranchENIs associated with the trunk
        uidToBranchENIMap map[string][]*ENIDetails
        // deleteQueue is the queue of ENIs that are being cooled down before being deleted
@@ -142,6 +144,7 @@ func NewTrunkENI(logger logr.Logger, instance ec2.EC2Instance, helper api.EC2API
                usedVlanIds:       availVlans,
                ec2ApiHelper:      helper,
                instance:          instance,
+               lastVlanIdx:       -1,
                uidToBranchENIMap: make(map[string][]*ENIDetails),
        }
 }
@@ -585,9 +588,11 @@ func (t *trunkENI) assignVlanId() (int, error) {
        t.lock.Lock()
        defer t.lock.Unlock()
 
-       for index, used := range t.usedVlanIds {
-               if !used {
+       for i := 0; i < len(t.usedVlanIds); i++ {
+               index := (t.lastVlanIdx + 1 + i) % len(t.usedVlanIds)
+               if !t.usedVlanIds[index] {
                        t.usedVlanIds[index] = true
+                       t.lastVlanIdx = index
                        return index, nil
                }
        }
```
(The code might be cleaner with `nextVlanIdx` instead, which would allow 0-initializing it)  

 <details>
  <summary>
 A couple of tests would also have to be updated.</summary>

```diff
diff --git a/pkg/provider/branch/trunk/trunk_test.go b/pkg/provider/branch/trunk/trunk_test.go
index 01a63ad..3f15abc 100644
--- a/pkg/provider/branch/trunk/trunk_test.go
+++ b/pkg/provider/branch/trunk/trunk_test.go
@@ -207,6 +207,7 @@ func getMockTrunk() trunkENI {
        log := zap.New(zap.UseDevMode(true)).WithName("node manager")
        return trunkENI{
                log:               log,
+               lastVlanIdx:       -1,
                usedVlanIds:       make([]bool, MaxAllocatableVlanIds),
                uidToBranchENIMap: map[string][]*ENIDetails{},
        }
@@ -233,7 +234,7 @@ func TestTrunkENI_assignVlanId(t *testing.T) {
        assert.NotNil(t, err)
 }
 
-// TestTrunkENI_freeVlanId tests if a vlan id is freed it can be re assigned
+// TestTrunkENI_freeVlanId tests if a vlan id is freed it is not assigned again on next allocation
 func TestTrunkENI_freeVlanId(t *testing.T) {
        trunkENI := getMockTrunk()
 
@@ -243,12 +244,12 @@ func TestTrunkENI_freeVlanId(t *testing.T) {
        assert.Equal(t, 0, id)
 
        // Free the vlan Id
-       trunkENI.freeVlanId(0)
+       trunkENI.freeVlanId(id)
 
        // Assign single Vlan Id again
-       id, err = trunkENI.assignVlanId()
+       id2, err := trunkENI.assignVlanId()
        assert.NoError(t, err)
-       assert.Equal(t, 0, id)
+       assert.NotEqual(t, id, id2)
 }
 
 func TestTrunkENI_markVlanAssigned(t *testing.T) {
```

</details>

Because we cannot deploy this code ourselves I have not tested it, and I obviously don't know this code, so maybe it's not as easy as it seems :)

PS! In case anyone else hits this and tries to search here: The behavior we observed was that all traffic was stuck in `SYN_SENT`.

**Comment by JacobHenner** [comment]
@SaranBalaji90 Do you have an estimate as to when a workaround, short-term fix, or long term fix might be available? This is affecting pods in our production clusters.

**Comment by chlunde** [comment]
PS! Prometheus runs with `terminationGracePeriodSeconds: 600` so it does not look like that is a mitigation of this issue as far as I can see.

**Comment by SaranBalaji90** [comment]
You are right @chlunde terminationGracePeriod setting doesn't help with this and yes I was talking about long term fix which team is actively looking into. I'm looking at other short-term fix to mitigate the impact. Will keep you posted here by EOD.

**Comment by SaranBalaji90** [comment]
@JacobHenner as per @jayanthvn who is driving this repo, next version release is scheduled in two weeks and we should have RC image before that. We will work on getting this fixed in the next release. If anything slips will keep this thread updated.

**Comment by jayanthvn** [comment]
@JacobHenner as Balaji mentioned we will be working on the file based solution. We are targeting for the fix in next couple weeks. Once we have the fix, we will cut a RC image and unblock you with the testing while we cut official release on GitHub. Hope this helps.

**Comment by chlunde** [comment]
@jayanthvn Thanks for the info. I want to verify that the PR fixes this issue, should the CNI driver work when I compile the main and init image from the PR, or are there other dependencies/known issues with that branch?

**Comment by jayanthvn** [comment]
@chlunde - Thanks for verifying. Yes you should be able to build the image with the PR based of master. There are no dependencies or known issues.

/cc @cgchinmay

**Comment by cgchinmay** [comment]
> @jayanthvn Thanks for the info. I want to verify that the PR fixes this issue, should the CNI driver work when I compile the main and init image from the PR, or are there other dependencies/known issues with that branch?

you should be able to build it without any other dependencies. 
```
make docker
```
and then use the generated cni image. And then update your [daemonset ](https://github.com/aws/amazon-vpc-cni-k8s/blob/6a15a84cae6576cba1c3302964431a26c435d07c/config/v1.9/aws-k8s-cni.yaml#L140)with this image and apply
```
kubectl apply -f https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/master/config/v1.9/aws-k8s-cni.yaml
```

**Comment by chlunde** [comment]
@cgchinmay @jayanthvn sadly I don't see an improvement with this version.
We have deployed kube-prometheus-stack w/ helm and persistent storage. There's a securitygrouppolicy in the namespace.

```yaml
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: kube-prometheus-stack
  namespace: monitoring
spec:
  chart:
    spec:
      chart: kube-prometheus-stack
      version: "19.1.0"
      sourceRef:
        kind: HelmRepository
        name: prometheus-community
        namespace: monitoring
  interval: 5m
  upgrade:
    crds: CreateReplace
  install:
    crds: CreateReplace
    remediation:
      retries: 3
  # Default values
  # https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/values.yaml
  values:
    kubeTargetVersionOverride: 1.21.2
    kubeProxy:
      enabled: false
    prometheus:
      prometheusSpec:
        image:
          repository: quay.io/prometheus/prometheus
          tag: v2.30.3 # TODO: dependencies.yaml
        replicas: 2
        resources:
          limits:
            cpu: 2000m
            memory: 8Gi
          requests:
            cpu: 500m
            memory: 8Gi
        storageSpec:
          volumeClaimTemplate:
            spec:
              storageClassName: gp2
              accessModes: ["ReadWriteOnce"]
              resources:
                requests:
                  storage: 50Gi
                  
        podMonitorSelectorNilUsesHelmValues: false
        serviceMonitorSelectorNilUsesHelmValues: false
        ruleSelectorNilUsesHelmValues: false
```

```
prometheus-kube-prometheus-stack-prometheus-1               2/2     Running   0          3m35s  
```
delete pod to trigger restart on same node:

```
k delete po -n monitoring prometheus-kube-prometheus-stack-prometheus-1
pod "prometheus-kube-prometheus-stack-prometheus-1" deleted
```

new pod is not `Ready` (`1/2` instead of `2/2`):
```
prometheus-kube-prometheus-stack-prometheus-1               1/2     Running   0          103s 
```

checking the logs of the new pod, we can see it does not reach the control plane service:

```
prometheus-kube-prometheus-stack-prometheus-1 prometheus level=error ts=2021-11-04T21:21:23.077Z caller=klog.go:116 component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.22.1/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: Get \"https://172.20.0.1:443/api/v1/namespaces/monitoring/pods?limit=500&resourceVersion=0\": dial tcp 172.20.0.1:443: i/o timeout"
```

Checking interfaces on the node:
```
sh-5.1# ip a |grep vlan
11: vlan3ae77ebeea7@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
12: vlan.eth.2@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
14: vlan0996cc66d36@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
15: vlan.eth.3@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
16: vlan3516bb1ebe7@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
17: vlan.eth.4@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
20: vlaneca9ff6a09a@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
21: vlan.eth.5@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
22: vlan88b786b9a3c@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
23: vlan.eth.6@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
26: vlan67eda70f998@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
27: vlan.eth.7@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
28: vlanf40a282689b@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
29: vlan.eth.9@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
31: vlan9201def753a@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
32: vlan.eth.8@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
33: vlanc83fc0c8c96@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
34: vlan.eth.10@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
37: vlan58f92f62c84@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default 
sh-5.1# 
```
one `vlan.eth.X iface` missing, (more vlanHASH interfaces than vlan.eth.x)

`kubectl describe pod` shows:

```
  Normal   ResourceAllocated       3m49s                   vpc-resource-controller  Allocated [{"eniId":"eni-06fcaa0c7d84c3037","ifAddress":"06:f3x","privateIp":"1c","vlanId":11,"subnetCidr":"x/24"}] to the pod
```
`vlan.eth.11` missing above

**Comment by cgchinmay** [comment]
@chlunde Could you email ipamd.log and plugin.log from this impacted node to cgadgil@amazon.com ?

**Comment by chlunde** [comment]
Logs sent. Yes, your sequence is correct.

**Comment by cgchinmay** [comment]
> Logs sent. Yes, your sequence is correct.

I looked at the logs and could not find any entry for Pod: prometheus-kube-prometheus-stack-prometheus-1 . Could you run eks-log-collector and send entire zip file to cgadgil@amazon.com .

**Comment by chlunde** [comment]
> > Logs sent. Yes, your sequence is correct.
> 
> 
> 
> I looked at the logs and could not find any entry for Pod: prometheus-kube-prometheus-stack-prometheus-1 . Could you run eks-log-collector and send entire zip file to cgadgil@amazon.com . 

Sorry for the confusion, the node was terminated by the autoscaler so I reproduced the issue again. As the kubectl commands in the mail show, this time the affected pod was -0 instead of -1.

I'll do some more testing when I get to work today and see if I can run the log collector command on bottlerocket if you need the additional logs it collects.

**Comment by cgchinmay** [comment]
> > > Logs sent. Yes, your sequence is correct.
> > 
> > 
> > I looked at the logs and could not find any entry for Pod: prometheus-kube-prometheus-stack-prometheus-1 . Could you run eks-log-collector and send entire zip file to [cgadgil@amazon.com](mailto:cgadgil@amazon.com) .
> 
> Sorry for the confusion, the node was terminated by the autoscaler so I reproduced the issue again. As the kubectl commands in the mail show, this time the affected pod was -0 instead of -1.
> 
> I'll do some more testing when I get to work today and see if I can run the log collector command on bottlerocket if you need the additional logs it collects.

I am checking the logs, will update my findings. Meanwhile do share the entire logs using eks-log-collector

**Comment by cgchinmay** [comment]
I checked earlier logs and here's the update
1. I found successful Create and Delete for 4 containers in this pod : prometheus-kube-prometheus-stack-prometheus-0
Not sure why Delete is triggered, something that we need to investigate so need to have entire logs
2. I could not find any entry for prometheus-kube-prometheus-stack-prometheus-0 using vland: 11 as seen in your output. 

I did find some gaps in our logging, so I will update the PR accordingly. 
My suspicions is that these repeated create and delete container calls could be blocking the pod from reaching a ready state. Need to understand if this is related to container runtime, so I will test with containerd runtime as in your case.
Also, if you can repro the issue and send me the entire logs, that would be very helpful.

**Comment by chlunde** [comment]
@cgchinmay nice work! I only noticed the last delete event, which was why I was worried about race conditions and overlapping use of the same VLAN and potential slowness of pod starts after restart. I notice that bottlerocket tests are run weekly, so I hope that will prevent regressions.

I've run the updated PR and done quite a few restarts, I have not yet run into any issues (except the unrelated 1727)!

**Comment by cgchinmay** [comment]
> @cgchinmay nice work! I only noticed the last delete event, which was why I was worried about race conditions and overlapping use of the same VLAN and potential slowness of pod starts after restart. I notice that bottlerocket tests are run weekly, so I hope that will prevent regressions.
> 
> I've run the updated PR and done quite a few restarts, I have not yet run into any issues (except the unrelated 1727)!

Thanks for confirming.

**Comment by chlunde** [comment]
@cgchinmay do you have a status update?

**Comment by cgchinmay** [comment]
> @cgchinmay do you have a status update?

@chlunde By status update if you mean official public image release(v.1.10.2) for the above fix then its scheduled for 7th Feb. But we do have a private image released with the above fix (v.10.2-rc1) and have been asking Customers to try it out until then. 

Installation instructions for private image : v1.10.2-rc1
Apply this [manifest](https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/v1.10.1/config/master/aws-k8s-cni.yaml) with the [cni image tag](https://github.com/aws/amazon-vpc-cni-k8s/blob/104354c45bb210c444bb3317e4199b696cfbb2f9/config/master/aws-k8s-cni.yaml#L142) changed to v1.10.2-rc1

**Comment by cgchinmay** [comment]
Closing this issue as we have released v1.10.2 with fix. Feel free to reopen the issue if you encounter problems.
